### PR TITLE
SIP-14 backport to 2.9.x

### PR DESCRIPTION
**Pull request by adriaanm**
Other than adding the new APIs, a few changes are required:
1. Build. The Java sources in "scala/concurrent/impl" need forkjoin.jar on the classpath. Thus, I replaced the "classpath" attribute in the respective `javac` ant tasks (locker.lib and quick.lib) with a "classpathref" attribute pointing to a classpath which includes also the forkjoin.jar. In the locker.lib target, exclude the duration package and everything that depends on it.
   
   Also, the library is built with -Ydependent-method-types (for both quick & strap -- use consistent options for stability)
2. Dependent method types and SI-5958. The `duration` package relies on dependent method types, and requires a backport of SI-5958 (original backport was in #1726).
3. To avoid a backend bug involving `throw` expressions inside closures (exercised by the Try tests), a backport of the fix for SI-6188 is required (original backport was  in #1755).

Review by @jsuereth (build)

(resubmission of https://github.com/scala/scala/pull/1745)

### CODE REVIEW

File: build.xml [hunk 1]
```diff
@@ -288,7 +288,8 @@ INITIALISATION
        This is to facilitate testing new command line options which do not yet exist in starr. -->
   <property name="scalac.args.quickonly" value=""/>
     <property name="scalac.args.all" value="${scalac.args} ${scalac.args.optimise}"/>
-	  <property name="scalac.args.quick" value="${scalac.args.all} ${scalac.args.quickonly}"/>
+          <property name="scalac.args.quick" value="${scalac.args.all} ${scalac.args.quickonly}"/>
+    <property name="scalac.args.strap.lib" value="${scalac.args.all} -Ydependent-method-types"/>
```
Comment (jsuereth) [inline build.xml]: stap should be using the same args as "quick", otherwise odd things could happen....

File: build.xml [hunk 2]
```diff
@@ -563,7 +576,7 @@ QUICK BUILD (QUICK)
       destdir="${build-quick.dir}/classes/library"
       compilerpathref="locker.classpath"
       srcpath="${src.dir}/library"
-      params="${scalac.args.quick}"
+      params="${scalac.args.quick} -Ydependent-method-types"
```
Comment (jsuereth) [inline build.xml]: Let's get these into the default scalac.args.quick

File: build.xml [hunk 3]
```diff
@@ -1100,7 +1113,7 @@ BOOTSTRAPPING BUILD (STRAP)
       destdir="${build-strap.dir}/classes/library"
       compilerpathref="pack.classpath"
       srcpath="${src.dir}/library"
-      params="${scalac.args.all}"
+      params="${scalac.args.strap.lib}"
```
Comment (jsuereth) [inline build.xml]: I had a fix in 2.10 somewhere that changes this one to be saclac.args.quick, as it stands it's kind of bad.

**Comment by adriaanm** [comment]
PLS REBUILD pr-scala-testsuite-linux-opt

**Comment by jsuereth** [comment]
LGTM

**Comment by adriaanm** [comment]
PLS REBUILD pr-scala-testsuite-linux-opt

**Pull request merged**